text/0003-templates.md

```diff
@@ -0,0 +1,258 @@
+  "pnpm": {
+    "templates": {
+      "extends": ["@example/react-template@0.1.0"],
+    }
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "version": "0.1.0",
+  "author": "Example Team <team@team.example>",
+  "license": "MIT",
+  "scripts": {
+    "compile": "tsc"
+  },
+  "dependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+  }
+}
+```
+
+Note that non-standard `package.json` fields such as `eslintConfig` will also be copied. Although these fields have no meaning to package managers, they will appear in the published manifest.
+
+### Toolkit
+
+The `pnpm.templates.toolkit` field copies the `dependencies` block of the template and spreads it into the `devDependencies` of the referencing package.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "toolkit": ["@example/react-template@0.1.0"],
+    }
+  },
+  "devDependencies": {
+    "jest": "^29.4.3"
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "dependencies": {
+    "redux": "^4.2.1"
+  },
+  "devDependencies": {
+    "jest": "^29.4.3",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+  }
+}
+```
+
+This allows package authors to share a reusable set of tools that can be imported or required in the referencing package. Dependencies of dependencies are not normally visible in this manner since pnpm sets up a semi-strict `node_modules` structure by default.
+
+### Catalog
+
+The `pnpm.templates.catalog` feature allows packages to declare a dependency using a version specifier from the referenced template. The `dependencies` block of the template will be available to reference through the `catalog:` version specifier protocol.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "catalog": ["@example/react-template@0.1.0"],
+    }
+  },
+  "dependencies": {
+    "react": "catalog:"
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "dependencies": {
+    "react": "^18.2.0"
+  }
+}
+```
+
+We expect most monorepos to use this feature to keep dependency specifiers consistent between different in-repo packages.
+
+### Combining Templates
+
+Templates may not refer to other templates. Instead of an inheritance hierarchy, a package may refer to multiple templates with later entries in the list taking precedence.
+
+For example, if both `@organization/authoring-metadata` and `@team/authoring-metadata` have an `author` field, the value from `@team/authoring-metadata` will be used.
+
+```json5
+{
+  "name": "@example/simple",
+  "pnpm": {
+    "templates": {
+      "extends": [
+        "@organization/authoring-metadata@0.1.0",
+        "@team/authoring-metadata@0.1.0"
+      ],
+    }
+  }
+}
+```
+
+Avoiding inheritance simplifies implementation and usability in several ways.
```

Comment:
Previous discussion on inheriting multiple: https://github.com/pnpm/pnpm/issues/2713#issuecomment-1397647979

> I don't know if it makes sense to support multiple. This is inspired by Bit, which supports one env per component. I think one env should be enough because you can create envs that inherit other envs.

Single inheritance was proposed here, but I could see it being a common case to use multiple templates from different authors.

Reply:
I agree that templates shouldn't be allowed to extend templates. Let's keep this simple. Composing _might_ introduce some minor repetition but it keeps the subsystem clean.

Reply:
Cool. Sounds like we're in agreement that full-on inheritance is messy and simple composition is preferable. Thanks for the thoughts.